cpp/include/cudf/io/text/detail/tile_state.hpp

```diff
@@ -0,0 +1,174 @@
+
+#pragma once
+
+#include <cub/block/block_scan.cuh>
+
+namespace cudf {
+namespace io {
+namespace text {
+namespace detail {
+
+enum class scan_tile_status : uint8_t {
+  oob,
+  invalid,
+  partial,
+  inclusive,
+};
+
+template <typename T>
+struct scan_tile_state_view {
+  uint64_t num_tiles;
+  scan_tile_status* tile_status;
+  T* tile_partial;
+  T* tile_inclusive;
+
+  __device__ inline void initialize_status(cudf::size_type base_tile_idx,
+                                           cudf::size_type count,
+                                           scan_tile_status status)
+  {
+    auto thread_idx = blockIdx.x * blockDim.x + threadIdx.x;
+    if (thread_idx < count) {  //
+      // this is UB if tile_status gets assigned from multiple threads.
+      tile_status[(base_tile_idx + thread_idx) % num_tiles] = status;
+    }
+  }
+
+  __device__ inline void set_partial_prefix(cudf::size_type tile_idx, T value)
+  {
+    auto const offset = (tile_idx + num_tiles) % num_tiles;
+    cub::ThreadStore<cub::STORE_CG>(tile_partial + offset, value);
+    __threadfence();
+    cub::ThreadStore<cub::STORE_CG>(tile_status + offset, scan_tile_status::partial);
+  }
+
+  __device__ inline void set_inclusive_prefix(cudf::size_type tile_idx, T value)
+  {
+    auto const offset = (tile_idx + num_tiles) % num_tiles;
+    cub::ThreadStore<cub::STORE_CG>(tile_inclusive + offset, value);
+    __threadfence();
+    cub::ThreadStore<cub::STORE_CG>(tile_status + offset, scan_tile_status::inclusive);
+  }
+
+  __device__ inline T get_prefix(cudf::size_type tile_idx, scan_tile_status& status)
+  {
+    auto const offset = (tile_idx + num_tiles) % num_tiles;
+
+    while ((status = cub::ThreadLoad<cub::LOAD_CG>(tile_status + offset)) ==
+           scan_tile_status::invalid) {
+      __threadfence();
+    }
+
+    if (status == scan_tile_status::partial) {
+      return cub::ThreadLoad<cub::LOAD_CG>(tile_partial + offset);
+    } else {
+      return cub::ThreadLoad<cub::LOAD_CG>(tile_inclusive + offset);
+    }
+  }
+
+  __device__ inline T get_inclusive_prefix(cudf::size_type tile_idx)
```

Comment:
We're not using this currently, right? This is the version which would "busy poll" this one tile until its `inclusive` prefix is ready, correct?

Reply:
good catch. we're grabbing the final inclusive prefix from the host side, so this function isn't used.